proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,483 @@
+
+The following expressions are lex errors, according to the above grammar:
+
+* ``s"a ${} b"``
+
+  * Expression is missing
+
+* ``s"a ${=} b"``
+
+  * Not an expression
+
+* ``s"a ${let x =} b"``
+
+  * Incomplete expression
+
+* ``s"a ${{b} c"``
+
+  * The second ``{`` is not a valid character to start an expression
+
+See the "Semantics" and "Examples" sections to see examples of valid interpolated strings.
+
+Machinery
+~~~~~~~~~
+
+An interpolated string expression desugars to calls to ``fromBuilder``, ``toBuilder``, and ``interpolate``, which are defined as:
+
+::
+
+  -- | Laws:
+  --     * fromBuilder . toBuilder === id
+  --     * toBuilder . fromBuilder === id
+  class Monoid (Builder s) => Buildable s where
+    type Builder s = r | r -> s
+    toBuilder :: s -> Builder s
+    fromBuilder :: Builder s -> s
+
+  class Buildable s => Interpolate a s where
+    interpolate :: a -> Builder s
+
+Instances for ``String`` will be defined as well:
+
+::
+
+  newtype StringBuilder = StringBuilder (Endo String)
+    deriving newtype (Semigroup, Monoid)
+
+  instance Buildable String where
+    type Builder String = StringBuilder
+    toBuilder s = StringBuilder (Endo (s ++))
+    fromBuilder (StringBuilder (Endo f)) = f []
+
+  instance Interpolate String String where
+    interpolate = toBuilder
+  instance {-# OVERLAPPABLE #-} Show a => Interpolate a String where
+    interpolate = StringBuilder . Endo . shows
+
+This design allows interpolating anything, even user-defined types, into a ``String`` with ``Show``, but can be overridden for specific types. See the "Examples" section for more details.
+
+These definitions would initially be implemented in ``ghc-experimental`` under ``Data.String.Interpolate``. After the API has stablized, these would eventually live in ``GHC.Exts`` alongside ``IsString``.
+
+Semantics
+~~~~~~~~~
+
+With the machinery defined above, the following interpolated string desugars to the below expression:
+
+::
+
+  -- original string
+  s"foo ${f a b} bar ${g x} baz ${name}"
+
+  -- desugared
+  fromBuilder $
+    toBuilder "foo "
+    <> interpolate (f a b) <> toBuilder " bar "
+    <> interpolate (g x)   <> toBuilder " baz "
+    <> interpolate name    <> toBuilder ""
+
+The string literals there will be handled by ``-XOverloadedStrings`` as usual, if enabled.
+
+Interpolated multiline strings should desugar into a single-line string `as usual <https://github.com/brandonchinn178/ghc-proposals/blob/master/proposals/0569-multiline-strings.rst>`_ first, before interpolating expressions.
+
+::
+
+  let x = "hello"
+
+  -- original string
+  let str0 =
+        s"""
+        ${x} world
+        world ${x}
+        ${x} world
+        """
+
+  -- resolve multiline string
+  let str1 = s"${x} world\nworld ${x}\n${x} world"
+
+  -- resolve interpolation
+  let str2 = "hello world\nworld hello\nhello world"
+
+Examples
+--------
```

Comment:
A thing I think is lacking in existing interpolation (but that IIRC python's f-strings do quite well), is the ability to control how a value is formatted. E.g. "this `Day` should be formatted like Thu, 8 Aug 2024" or "this `Float` should have 2 decimal places".

I don't think this proposal needs to enable that kind of thing, but is it worth briefly discussing how it might look given this proposal plus some other library? E.g. I'm imagining

    s"today is ${Fmt "%a, %d %b %Y" todaysDate}"
    s"finished in ${Fmt ".2" durationSecs} seconds"

where some library provides relevant instances for `Fmt`.

The point isn't to figure out exactly how this use case should be handled, but to show that there is at least some reasonable way of handling this use case.

Reply:
You can use use custom formatting functions in conjunction with the `&` operator: `s"today is ${date & fmt "%a, %d %b %Y"}`

Reply:
If you use a module with interpolable data types, Fmt is an unnecessary intermediate step, since you can have data types that do the equivalent of .2 etc: `s"finished in ${Fmt.Prec 2 durationSecs} seconds"`. This leaves less room for errors in specifiers or how they're used, or places where it doesn't make sense to interpolate with such specifiers.

Reply:
Added some examples

Reply:
Note that for floats, you might want to specify: total width; precision after decimal point; whether to have a `+` for positives or not; whether to left-pad with 0 or space; whether to use `,` or `_` or nothing to separate groups of digits; and whether to use `%f`, `%g` or `%e` style formatting.

So if I'm looking at this and wondering whether I'll be able to do what I want in a way that feels reasonable, the current example doesn't feel very compelling.

And in fact, with the current design, I think the answer is no: because I can't have a `"` inside `${ }`, I can't do what I want in a way that feels reasonable. Number formatting feels like it would be awkward, and date formatting seems completely off the table.

I think I'd be satisfied with examples showing me that I can do these two things, in a way that doesn't feel like I'm jumping through hoops:

* Render a `Day` with a specific date format like `%a, %d %b %Y`.
* Render a `Float` similarly to `printf "%+8.2f"`.

To be clear, the thing I'm looking for from examples here is:

* Here's a thing someone might reasonably want to do;
* Here's a way that might reasonably look, if someone implemented a library to do so.

I don't think you need to show how that library would be implemented, as long as it seems clear that it could be.